experimental/codepointtrie/src/codepointtrie.rs

```diff
@@ -0,0 +1,311 @@
+    const FAST_MAX: u32 = SMALL_TYPE_FAST_INDEXING_MAX;
+    const ENUM_VALUE: TrieTypeEnum = TrieTypeEnum::Small;
+}
+
+/// This struct represents a de-serialized CodePointTrie that was exported from
+/// ICU binary data.
+///
+/// For more information:
+/// - [ICU Site design doc](http://site.icu-project.org/design/struct/utrie)
+/// - [ICU User Guide section on Properties lookup](https://unicode-org.github.io/icu/userguide/strings/properties.html#lookup)
+pub struct CodePointTrie<'trie, W: ValueWidth, T: TrieType> {
+    header: CodePointTrieHeader,
+    index: ZeroVec<'trie, u16>,
+    data: ZeroVec<'trie, W>,
+    _marker_ty: PhantomData<T>,
+}
+
+/// This struct contains the fixed-length header fields of a [`CodePointTrie`].
+pub struct CodePointTrieHeader {
+    pub index_length: u32,
+    pub data_length: u32,
+    pub high_start: u32,
+    pub shifted12_high_start: u16,
+    pub index3_null_offset: u16,
+    pub data_null_offset: u32,
+    pub null_value: u32,
+}
+
+impl<'trie, W: ValueWidth, T: TrieType> CodePointTrie<'trie, W, T> {
+    /// Returns a new [`CodePointTrie`] backed by borrowed data for the `index`
+    /// array and `data` array, whose data values have width `W`, for a trie
+    /// type `T`.
+    pub fn try_new(
+        header: CodePointTrieHeader,
+        index: ZeroVec<'trie, u16>,
+        data: ZeroVec<'trie, W>,
+    ) -> Result<CodePointTrie<'trie, W, T>, Error> {
+        if header.data_length < ERROR_VALUE_NEG_DATA_OFFSET {
+            return Err(Error::FromDeserialized {
+                reason: "Data array must be large enough to contain error value",
+            });
+        }
+
+        if header.data_length < HIGH_VALUE_NEG_DATA_OFFSET {
+            return Err(Error::FromDeserialized {
+                reason:
+                    "Data array must be large enough to contain value for range highStart..U+10FFFF",
+            });
+        }
+
+        if index.len() as u32 != header.index_length {
+            return Err(Error::FromDeserialized {
+                reason: "Length of index array does not match corresponding header value",
+            });
+        }
+
+        if data.len() as u32 != header.data_length {
+            return Err(Error::FromDeserialized {
+                reason: "Length of data array does not match corresponding header value",
+            });
+        }
+
+        let trie: CodePointTrie<'trie, W, T> = CodePointTrie {
+            header,
+            index,
+            data,
+            _marker_ty: PhantomData,
+        };
+        Ok(trie)
+    }
+
+    /// Returns the position in the data array containing the trie's stored
+    /// error value.
+    fn trie_error_val_index(&self) -> u32 {
+        self.header.data_length - ERROR_VALUE_NEG_DATA_OFFSET
+    }
+
+    fn trie_internal_small_index(&self, code_point: u32) -> u32 {
+        let mut index1_pos: u32 = code_point >> SHIFT_1;
+        if T::ENUM_VALUE == TrieTypeEnum::Fast {
+            assert!(
```

Comment:
Issue, here and below: I'm not convinced that this `assert!` is guaranteed to be true, since one can construct a CodePointTrie from garbage data.  I suggest changing it to `debug_assert!`, and then just making sure no code paths cause a crash (you should be fine so long as you never `unwrap`).